debezium-core/src/main/java/io/debezium/pipeline/ChangeEventSourceCoordinator.java

```diff
@@ -58,6 +59,7 @@
 
     private volatile boolean running;
     private volatile StreamingChangeEventSource streamingSource;
+    private final ReentrantLock commitOffsetLock = new ReentrantLock();
```

Comment:
Couldn't we simply use `synchronized(streamingSource) { ... }`? Is there any advantage to using the lock?

Reply:
I'm less familiar with `synchronized` so you might be right but I added the lock to protect against the `streamingSource` getting set to null between https://github.com/debezium/debezium/pull/1596/files#diff-4155443cdda33d58a9cf05816667605bR145-R146

I saw this failure pop up as a NPE during a test.

It's my understanding that `synchronized(streamingSource) { ... }` would prevent multiple threads from accessing the block for the same object but I don't see how that would prevent the race condition of different threads acting on `streamingSource` between `ChangeEventSourceCoordinator#start` and `ChangeEventSourceCoordinator#commitOffset`

Reply:
Now that I've had my ☕ , I'm thinking a little more creatively! I could see pushing the access logic into a separate function might work, ie:

```
private static void accessStreamingSource(Consumer<StreamingChangeEventSource> consumer) {
    synchronized(streamingSource) {
        consumer.accept(streamingSource);
    }
}
``` 

Is this along the lines of what you mean?

Reply:
Essentially, yes. I don't think the consumer indirection would be needed, just use synchronized(streamingSource) { ... } wherever it's accessed.

Reply:
I'm a bit confused here. We really only have one thread at a time inside `start` so a synchronized block there wouldn't block another thread from entering `commitOffset` and potentially hitting a race condition, or is my understanding here off?

Reply:
Not sure whether I get the question right, but essentially *every* usage to `streamingSource` must be guarded by a synchronized block; this will ensure that only ever one of these methods will be in the critical section where it accesses that source.